src/System.Linq/src/System/Linq/Select.cs

```diff
@@ -795,6 +849,32 @@ public List<TResult> ToList()
                 return list;
             }
 
+            public HashSet<TResult> ToHashSet(IEqualityComparer<TResult> comparer)
+            {
+                int count = _source.GetCount(onlyIfCheap: true);
+
+                HashSet<TResult> hashSet;
+
+                switch (count)
+                {
+                    case -1:
+                        hashSet = new HashSet<TResult>(comparer);
+                        break;
+                    case 0:
+                        return new HashSet<TResult>(comparer);
+                    default:
+                        hashSet = new HashSet<TResult>(count, comparer);
```

Comment:
This is also one of the places where a lot of duplicates can cause excessive memory allocation. You can just initialize the `HashSet` without all the ceremony and add everything to it here-- the count won't help us.